### DISCUSSION TITLE
Dev mode hot reload loses database connection

### DISCUSSION THREAD

**Discussion by seejamescode** [vercel/next.js#26427] [Ideas]
# Bug report

## Describe the bug

Loving v9! Just ran into an issue where I can’t use an existing database connection with the new API routes.

## To Reproduce

`mongooseConnect.js`:
```js
const mongoose = require("mongoose").set("debug", true);
const connection = {};

module.exports = async () => {
  if (connection.isConnected) {
    // Using existing database connection
    return;
  }

  // Using new database connection
  const db = await mongoose.connect("mongodb://localhost:27017/my-app",
    { useCreateIndex: true, useFindAndModify: false, useNewUrlParser: true }
  );

  connection.isConnected = db.connections[0].readyState;
};
```

1. Require the above function into an API endpoint that uses the new `pages/api/` feature and requires
2. Run `next dev`
3. Ping the API endpoint and see the correct response
4. Edit `mongooseConnect.js` and save
5. Ping the API endpoint again

## Expected behavior

The response will be the same. Instead, I get the following error:

```
MongooseError: You can not `mongoose.connect()` multiple times while connected.
```

This is because `db` returns to `{}` with the HMR.

## System information

- OS: macOS
- Version of Next.js: 9.0.0

## Additional context

Is there a different way we should be handling serverless database connections or is this a bug?

**Comment by Timer** [comment]
👋 Hi! I believe this is expected unless you add custom `module.hot` handling to not fully replace this module instance.

I'll have one of our team members look into this to confirm, though.

**Reply by technillogue** [reply]
how do you add custom module.hot handling?

**Comment by seejamescode** [comment]
Thanks @Timer! I was able to add `mongooseConnect.js` to the ignore list with the `next.config.js`. However, that understandably does not save the dev experience when editing the api route file.

Is there any documentation for adding `module.hot` in Next.js? I was not able to use the condition directly in `mongooseConnect.js`. Do I use it in `next.config.js`?

**Comment by Timer** [comment]
@seejamescode under further review this isn't possible right now. We'll tag this as a feature request, sorry about that!

**Comment by prowebat** [comment]
@seejamescode:
i encountered a similiar problem.
easiest way was to save the database connection in a global variable:
so you could replace all instances of connection with eg. global.dbConnection:

**Comment by grubxy** [comment]
> @seejamescode:
> i encountered a similiar problem.
> easiest way was to save the database connection in a global variable:
> so you could replace all instances of connection with eg. global.dbConnection:
Hi @prowebat 
should I use singleton to solve this same problem?

**Comment by seejamescode** [comment]
Overall, my issue was reading blog posts about database connections in serverless setups too fast. I was calling `mongooseConnect()` inside of my api functions instead outside of them. I do not really encounter this issue as much now.

Feel free to keep the issue open if you think it would still be a beneficial feature request for others. Otherwise, I am fine with closing it for user error!

**Comment by vixeven** [comment]
@seejamescode Can you add please a [guide](https://zeit.co/guides) or an [example](https://github.com/zeit/next.js/tree/canary/examples) with your approach?

**Comment by katywings** [comment]
+1 ^.^

**Comment by vvo** [comment]
Hey there, using Next.js 9.3, in dev mode, when an API route is rebuilt, and this API route references db.js where you initiate your database connection, then it will reconnect (and previous connections are just kept alive somehow indefinitely).

When using knex/objection.js/node-postgres (trying to put keywords for SEO for people with similar issues) then you'll get the infamous "sorry, too many clients already".

@Timer how do you handle that on your side? Every Next.js project using a database connection should have the issue at some point.

I have not managed to find a way to express "Ignore this file for auto reload" when that file is a server side module (like lib/db.js) used by one of the API pages.

Thanks!

**Comment by vvo** [comment]
Update: I was able to ignore a file to be reloaded by Next.js on change by using https://nextjs.org/docs/api-reference/next.config.js/custom-webpack-config and:

```js
    webpackDevMiddleware(config) {
      config.watchOptions.ignored.push(path.join(__dirname, "lib/db.js"));
      return config;
    },
```

But the issue is still that connection between different routes is not reused, as each route re-runs the lib/db.js file which is confusing. In a regular express.js server every file require is cached. Weird!

**Comment by vvo** [comment]
Note: while I was able to reproduce this once, I tried again and was not able to get constant connection leaking so I guess we can just close this. There's definitely a bug somewhere, most probably not in Next.js though. If someone else ends up here let me know!

**Comment by vvo** [comment]
Ok I managed to pinpoint exactly the issue I have been facing with knex connection pooling, see here: https://github.com/knex/knex/issues/3788

**Comment by vvo** [comment]
@Timer Do you think it would be feasible to have an option to say "Do not clear require cache for this module" so that we can avoid bugs on some libraries in dev mode? Thanks!

**Comment by brandonweiss** [comment]
Oh man, this one really got me. I’ve been using Prisma and Next.js and could not for the life of me figure out why periodically my database would error saying there were too many connections. I [opened an issue on the Prisma client](https://github.com/prisma/prisma-client-js/issues/228) to figure it out. At some point I realized that the dev server was reloading the file where the Prisma client is instantiated, eventually creating multiple clients and thus too many connections.

I tried @vvo’s recommendation of adding the file to `config.watchOptions.ignored` but that doesn’t seem to work for me. If I throw a `console.log` in that file I can see it’s getting reloaded pretty much every time a file changes.

This is one of those edge-casey situations where technically neither the database library nor Next are doing anything wrong, but in combination you get really unexpected results. I suspect the only way to really this would be with an official way to make Next ignore a file in dev mode?

**Comment by huv1k** [comment]
You can do workaround like this meanwhile
```ts
let prisma: PrismaClient
if (process.env.NODE_ENV !== 'production') {
  if (!global.prisma) {
    global.prisma = new PrismaClient({
      debug: true,
    })
  }
  prisma = global.prisma
} else {
  prisma = new PrismaClient()
}
```

**Reply by hungcung2409** [reply]
In my case, this solution does not work anymore in Next 13+

**Reply by karlhorky** [reply]
@chimung you can check out my update using `globalThis` over here:

- https://github.com/vercel/next.js/discussions/26427#discussioncomment-898067

We have a demo app using this method that is using Next.js v15 (`next@15.0.0-canary.2`):

- https://github.com/upleveled/next-js-example-spring-2024-atvie/blob/34d5027a5670852b11c8812e0ac3c99cb0d933a4/database/connect.ts#L9-L41

**Reply by hungcung2409** [reply]
> @chimung you can check out my update using `globalThis` over here:
> 
> * [Dev mode hot reload loses database connection #26427 (comment)](https://github.com/vercel/next.js/discussions/26427#discussioncomment-898067)
> 
> We have a demo app using this method that is using Next.js v15 (`next@15.0.0-canary.2`):
> 
> * https://github.com/upleveled/next-js-example-spring-2024-atvie/blob/34d5027a5670852b11c8812e0ac3c99cb0d933a4/database/connect.ts#L9-L41

I have trouble with setting up NextJS v15. Do you have a repo for NextJS v14?

**Reply by karlhorky** [reply]
It's the same for v14, no changes were necessary for v15.

But if you want to see a v14 example, we have other example repos:

- https://github.com/orgs/upleveled/repositories?q=mirror%3Afalse+fork%3Afalse+archived%3Afalse+next-js-example+atvie

**Comment by brandonweiss** [comment]
@huv1k 🙌🏼🙌🏼🙌🏼 I completely forgot about `global`—thanks so much! ❤️

**Comment by vvo** [comment]
@huv1k you're the real MVP, I solved my issue this way too:

```js
const knex = require("knex");
let pg;

if (process.env.NODE_ENV === "development") {
  global.pg =
    global.pg ||
    knex({
      client: "pg",
      connection: process.env.DATABASE_URL,
    });

  pg = global.pg;
} else {
  pg = knex({
    client: "pg",
    connection: process.env.DATABASE_URL,
  });
}
```

**Comment by icflorescu** [comment]
Had the same issue for the past 10 months while working on a large project with Next.js + [Objection.js](https://vincit.github.io/objection.js/)/[Knex](http://knexjs.org/) + PostgreSQL.
Tried lots and lots of possible workarounds to no avail, until I found this thread and implemented the following hack based on @huv1k's suggestion:

```js
/**
 * @returns {import('knex')}
 */
const getConnection = () => {
  if (PRODUCTION) return knex(DB);
  /**
   * This `global` hack solves a long-standing issue appearing in development more,
   * when successive hot-reloads lead to connections being leaked and eventually produce
   * an unrecoverable error
   */
  /* eslint-disable no-underscore-dangle */
  if (!global.__KNEX_DB_CONNECTION__) global.__KNEX_DB_CONNECTION__ = knex(DB);
  return global.__KNEX_DB_CONNECTION__;
  /* eslint-enable */
};

const db = getConnection();

export default db;
```

Thanks a lot! 🙏🏻

**Comment by mattcarlotta** [comment]
~~I'm not so sure polluting the `global` space is the best work-around. This can be confusing as to where it's being defined (globally, but from what file?) and where it should be used (client? server?). An alternative approach I've found has been to utilize next's [runtime configs](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration) to prevent connection leaks in development. This also has the added benefit of hinting at where the function is being defined (from `next/config` which to points to the `next.config.js` file) and where it should be used (on the server).~~

On further testing, doesn't appear to work with `getStaticProps` and `getStaticPaths`. On the same note, the global example mentioned above still creates duplicate multiple connections in development on each page transition and creates `getStaticProps ` usage + dynamic page n connections in production.

Duplicate connections from navigating back and forth between index and the same dynamic page (3 iSSG pages):
![](https://i.imgur.com/1RRAQ4H.png)

**Comment by karlhorky** [comment]
I'm also having a similar issue with Next.js where the database connections of [Postgres.js](https://github.com/porsager/postgres) start to accumulate during a development workflow (page reloading, Fast Refresh, etc), leading to the dreaded PostgreSQL error:

```
Error: remaining connection slots are reserved for non-replication superuser connections
```

@huv1k is the `globalThis` option you mentioned above still the recommended workaround? Or is there a new way of getting around this in Fast Refresh? Would be great to have a hook into the refresh lifecycle to be able to terminate any open connections.

Here is our current workaround:

```tsx
import postgres from 'postgres';

declare module globalThis {
  let postgresSqlClient: ReturnType<typeof postgres> | undefined;
}

// Connect only once to the database
// https://github.com/vercel/next.js/discussions/26427#discussioncomment-898067
function connectOnceToDatabase() {
  if (!globalThis.postgresSqlClient) globalThis.postgresSqlClient = postgres();
  return globalThis.postgresSqlClient;
}

export const sql = connectOnceToDatabase();
```

Code: https://github.com/upleveled/next-js-example-fall-2022-vienna-austria/blob/d89b8243040c977b73c4bfae7519c60bced9090a/database/connect.ts

Update (June 2024):

- New repo using similar code with Next.js v15 (`next@15.0.0-canary.2`): https://github.com/upleveled/next-js-example-spring-2024-atvie/blob/34d5027a5670852b11c8812e0ac3c99cb0d933a4/database/connect.ts#L9-L41

**Reply by Eprince-hub** [reply]
This error/behavior is still happening in Next.js 13 with the `/app` directory

https://github.com/vercel/next.js/discussions/26427#discussioncomment-4796502

**Reply by nanzm** [reply]
globalThis will be reset in Next.js 13 with the /app directory

**Reply by karlhorky** [reply]
> globalThis will be reset in Next.js 13 with the /app directory

I'm not sure what you mean with this, but we've had no problems with Next.js 13 and 14 using the App Router:

https://github.com/upleveled/next-js-example-fall-2023-atvie/blob/49793411b24c12eb9d092026c7bb48332d8c699d/database/connect.ts#L12-L23

**Reply by nanzm** [reply]
I test your code. Every time I refresh the page, a new database connection will be created.


<img width="728" alt="image" src="https://github.com/vercel/next.js/assets/12454813/93e49408-8e98-499d-88bd-b85a4741b4b1">


![2222](https://github.com/vercel/next.js/assets/12454813/828708a7-5954-49fc-897c-a733f6c2ec15)

**Reply by Eprince-hub** [reply]
@nanzm, 
What I found so far about your report

- The calling of the `connectOneTimeToDatabase()` function when the database is already connected happens only on dynamic segments and even though it is being called on every reload, the connection slot count is not being increased, and I think that is why the connection slot error does not resurface. I haven't yet figured out why calling the function doesn't cause the connection count to increase though. 
 @porsager, Do you have any ideas why calling the `postgres()` function again wouldn't cause the connection count to increase?
https://github.com/upleveled/next-js-example-fall-2023-atvie/blob/49793411b24c12eb9d092026c7bb48332d8c699d/database/connect.ts#L12-L23

- When you connect to the database without using the `connectOneTimeToDatabase()` function method, you can notice the connection count increasing each time the page reloads. So I think the `connectOneTimeToDatabase()` is still a valid solution as shown in this thread.

Used connection count doesn't increase even as the function is repeatedly called

https://github.com/vercel/next.js/assets/74430629/c1e2d5e9-5786-4434-a193-892b54a81682

Connecting without the `connectOneTimeToDatabase()` function

https://github.com/vercel/next.js/assets/74430629/58c26b81-5d9d-402e-a46e-6b93e938b237

**Reply by Eprince-hub** [reply]
> globalThis will be reset in Next.js 13 with the /app directory

@nanzm Do you have any official docs where this statement was made?

Indeed this behavior of reconnecting to the database from a dynamic route despite the usage of the `connectOneTimeToDatabase()` happens only on projects using the `next.js 13+` with the app directory. Projects before that work fine with the `connectOneTimeToDatabase()`  and connect only once to the database.  https://github.com/upleveled/graphql-example-fall-2022-atvie/blob/7d9a8f1bb725e5498eecb44a7abb764cccfbe191/database/connect.ts#L12-L22

**Reply by Eprince-hub** [reply]
The workaround in this thread doesn't prevent the database connection from reconnecting on every reload on dynamic routes/segments, even though the `connectOnceToDatabase()` workaround is still valid because it still prevents the accumulation of the connection counts that cause the `Error: remaining connection slots are reserved for non-replication superuser connections` error in [Postgres.js](https://github.com/porsager/postgres/tree/master)

https://github.com/vercel/next.js/assets/74430629/c1e2d5e9-5786-4434-a193-892b54a81682

I don't know why the subsequent connection to the database doesn't increase the count but I found out that these subsequent connections don't go deep past this [Sql function](https://github.com/porsager/postgres/blob/6f20a4820c683b33e7670b606d8daf5670f4b973/src/index.js#L86) in `Postgres.js` 

These functions get called only once when the initial connection happens and don't get called on subsequent connections🤔 

https://github.com/porsager/postgres/blob/6f20a4820c683b33e7670b606d8daf5670f4b973/src/connection.js#L326
https://github.com/porsager/postgres/blob/6f20a4820c683b33e7670b606d8daf5670f4b973/src/connection.js#L352
https://github.com/porsager/postgres/blob/6f20a4820c683b33e7670b606d8daf5670f4b973/src/connection.js#L356

**Could this explain why the connection count doesn't accumulate?**
**Maybe the subsequent connections aren't happening?**

I have no other explanation

![Screenshot 2023-12-05 at 15 27 29](https://github.com/vercel/next.js/assets/74430629/7ab1de2a-6400-4c5d-b8a3-ec1220639a2e)

![Screenshot 2023-12-05 at 15 27 42](https://github.com/vercel/next.js/assets/74430629/2f8b6252-91a1-4ab2-92ac-60f0b39dddd4)

https://github.com/vercel/next.js/assets/74430629/ba2a9384-4551-4aa9-8fdb-a88f259b822e

**Reply by nanzm** [reply]
Sorry for not replying to you in a timely manner. Thank you for answering my question. Postgres.js may have some special mechanism to prevent the creation of a large number of connections within a short period of time. 

In my project, I’m using Prisma to connect to a MySQL database. Every time I refresh the page, a new connection is created, and I quickly get the error “Too many connections.” 

From my observation, this only happens in development mode because globalThis is reset every time the page is refreshed. Therefore, implementing a singleton pattern is not feasible in this case, and I haven’t found a solution yet.

**Comment by karlhorky** [comment]
@huv1k Actually, would the Next.js team be open to a new page in the docs about "Database Connections and Other Side Effects"?

I would be open to writing a short guide on how to create singletons for this using `globalThis`, if this is the current official recommendation.

CC @Timer @timneutkens

**Comment by noclat** [comment]
A bit late here but you guys all saved my life. I'm using Bull and each API call would create new Redis clients, so I had to restart the server frequently (like every 10 minutes) when developing because of memory leaks.

I package all my instances in a `src/server/services/index.js` file where I init databases, Bull and other third-party services (SDKs). So I've extracted the `global` logic into a helper:

```js
const registerService = (name, initFn) => {
  if (process.env.NODE_ENV === 'development') {
    if (!(name in global)) {
      global[name] = initFn();
    }
    return global[name];
  }
  return initFn();
};
```

So that I can create unique instances of my services by simply calling:

```js
// src/server/services/index.js
export const db = registerService('db', () => knex({
  client: 'pg',
  connection: process.env.DATABASE_URL,
}));

// src/pages/api/...
import { db } from 'server/services';
```

for example.

I've spent so many hours trying to find a way to fix that. And this works perfectly. Thanks a lot!

**Comment by jeff-kilbride** [comment]
For knex / objection, I've been setting things up in the `next.config.js` file:

```javascript
const Knex = require('knex');
const { Model } = require('objection');
const {
  PHASE_PRODUCTION_SERVER,
  PHASE_DEVELOPMENT_SERVER,
} = require('next/constants');

const knexConfig = require('./knexfile.js');

const serverPhases = [PHASE_DEVELOPMENT_SERVER, PHASE_PRODUCTION_SERVER];

module.exports = (phase) => {
  if (serverPhases.includes(phase)) {
    const knex = Knex(knexConfig);
    Model.knex(knex);
    console.log('Initialized objection models...');
  }

  return {
    reactStrictMode: true,
  };
};
```
with `knexfile.js` like this:
```javascript
module.exports = {
  client: 'mysql',
  connection: {
    host: process.env.DB_HOST,
    port: process.env.DB_PORT,
    user: process.env.DB_USER,
    password: process.env.DB_PASS,
    database: process.env.DB_NAME,
  },
  migrations: {
    directory: './src/models/migrations',
    tableName: 'knex_migrations',
    stub: './src/models/migrations/migration.stub',
  },
};
```
This works well for me, but I haven't been hammering my dev environment. So, I haven't had an occasion where I've run out of connections. I do get two `'Initialized objection models...'` when I start the dev server, though.

I'd be interested if this worked for someone else -- or if there was a reason why I **shouldn't** do it this way!

**Comment by technillogue** [comment]
it would be really great if there was a way to tell next to not clear a module from require.cache - per https://github.com/knex/knex/issues/3788#issuecomment-613996800 something is possible with NextJsRequireCacheHotReloader, but I'm not sure how to actually override it.

**Comment by Eprince-hub** [comment]
The new Next.js 13 with the `/app` directory still has this problem of the database connections of [Postgres.js](https://github.com/porsager/postgres) accumulating during development that leads to the PostgreSQL error below:

```
remaining connection slots are reserved for non-replication superuser connections
```


![Screenshot 2023-01-26 at 16 31 17](https://user-images.githubusercontent.com/74430629/215083910-5ae6c3cd-b5c0-426a-b539-3f509b6a74a4.png)

During development, because of the Fast Refresh (hot reloading), the database connection count increases and causes the error when it reaches 90 and above

I have the database connection in the `database.ts` file and a code that exports a database query function like the example below
```js
const sql = postgres();
```

```js
export async function getFruits() {
  const fruits = await sql`
    SELECT * FROM foods
  `;
  return fruits;
}
```

Importing the database query function `getFruits` into any server component in the `app` directory, e.g: `app/layout.ts` causes Fast Refresh whenever the file is edited/saved, which causes this error

```ts
import { getFruits } from './database';

export default async function RootLout() {
  const fruits = await getFruits();

  return (
    <div>
      {fruits.map((fruit) => (
        <div key={fruit.id}>
          <h1>{fruit.name}</h1>
          <p>{fruit.description}</p>
        </div>
      ))}
    </div>
  );
}
```

The solution still remains the `globalThis` workaround that @karlhorky mentioned [here](https://github.com/vercel/next.js/discussions/26427#discussioncomment-898067)

**Comment by grrowl** [comment]
The server-side process on hot-reload receives a SIGTERM you can use to clean up connections. Wouldn't call this solution _ideal_ as it prevents work in progress during shutdown from writing to database; but it works, so lets get into it:

```ts
import postgres from "postgres";
import { drizzle } from "drizzle-orm/postgres-js";

const connection = postgres(connectionOptions);
export const db = drizzle(connection, { schema, logger: DB_LOGGING });

// Clean up connections on hot reload or process exit

process.on("SIGTERM", async () => {
  console.log("Database client disconnecting (SIGTERM)");
  await connection.end();
});

process.on("SIGINT", async () => {
  console.log("Database client disconnecting (SIGINT)");
  await connection.end();
});
```

If you're using a client that explicitly requires `client.close()`, like Redis, this is the only way to avoid hanging processes whether prod or dev.
